discord.go

```diff
@@ -0,0 +1,272 @@
+package panicbot
+
+import (
+	"fmt"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/bwmarrin/discordgo"
+)
+
+type Discord interface {
+	BanUser(userID string, reason string, days int) (discordgo.GuildBan, error)
+	SendChannelMessage(channelID string, message string) (*discordgo.Message, error)
+	SendDM(userID string, message string) (*discordgo.Message, error)
+}
+
+func (Discord *DiscordImpl) BanUser(userID string, reason string, days int) (discordgo.GuildBan, error) {
+
+	err := Discord.session.GuildBanCreateWithReason(Discord.guildID, userID, reason, days)
+	if err != nil {
+		return discordgo.GuildBan{}, fmt.Errorf("failed to ban user with userID:  %s", userID)
+	}
+
+	guildBan, err := Discord.session.GuildBan(Discord.guildID, userID)
+	if err != nil {
+		return discordgo.GuildBan{}, fmt.Errorf("failed to retrieve ban information for user with userID: %s", userID)
+	}
+
+	Discord.logger.WithFields(log.Fields{
+		"user":     guildBan.User.String(),
+		"reason":   reason,
+		"dateTime": time.Now().String(),
+	})
+
+	return *guildBan, nil
+}
+
+func (Discord *DiscordImpl) SendChannelMessage(channelID string, content string) (*discordgo.Message, error) {
+	message, err := Discord.session.ChannelMessageSend(channelID, content)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send message to primary channel")
+	}
+	Discord.logger.WithFields(log.Fields{
+		"author":    message.Author,
+		"channelID": message.ChannelID,
+		"guildID":   message.GuildID,
+		"message":   message.Content,
+		"dateTime":  time.Now().String(),
+	})
+	return message, nil
+}
+func (Discord *DiscordImpl) SendDM(userID string, message string) (*discordgo.Message, error) {
+	channel, err := Discord.session.UserChannelCreate(userID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create private message channel with userID: %s", userID)
+	}
+	return Discord.SendChannelMessage(channel.ID, message)
+}
+
+func PanicAlertCallback(s *discordgo.Session, i *discordgo.InteractionCreate) {
+	// TODO write logic for starting a panicalert vote
```

Comment:
These todos are valid,  but the issue is that the reason we made the call backs is so that this work would happen _not here_  But in the main package where it has action to container. (methods on container to be specific) 

It's also possible that we want to simplify the callbacks to not be identical to the callback that discord go uses and only give it the data it needs.  Handle the messing with the interaction create in the panic bot package etc. It would mean the helper down below would get a bit more comicated,  but it would likely be worth it.

Reply:
I would like to spend some time pair programming with you on this ticket, because I am not sure I fully understand the design pattern we implemented.

I understand how to pass a function to another function, but the calling code is where I am stumbling.

Reply:
The goal of using an interface is to abstract the complicated, repeated parts that we "dont care about" (IE actually interacting with discordgo). The "we" in this case, being the main container. Therefore, we (the programmers) created the discord interface which sets many things up discord side for us with regards to the slash commands, the bot joining, and the parsing the data we care about from the different events.  

The interface also then has a series of commands we can call IE `BanUser(userID string, reason string, days int)` with values the main container knows but discordgo would not understand. The interface implementation handles all the conversion from easy to use values to complicated discordgo structs, along with sending those structs TO discord.

As part of initialization, we pass all the information it needs to set this stuff up, along with methods on the main container for each of the call-backs. These methods will only take in the information we care about/need on the container side (such as UserID and reason from the panic-ban command as strings.). 

We then write a handler on the discord.go side that parses the command called, based on that command, parses the required information for , and finally calls the callback we passed in with only the required information.